### ISSUE TITLE
custom PPO loss does not work

### ISSUE THREAD

**Issue by cpn** [ray-project/ray#7855]
The following code which changes the default ppo loss works fine, despite having an assert(False) in the new loss. I believe the update does not happen. 
```
import logging

from ray.rllib.evaluation.postprocessing import Postprocessing
from ray.rllib.policy.sample_batch import SampleBatch
from ray.rllib.agents.ppo.ppo_tf_policy import PPOTFPolicy
from ray.rllib.utils import try_import_tf
from ray.rllib.models import ModelCatalog
tf = try_import_tf()

logger = logging.getLogger(__name__)


class CustomPPOLoss:
    def __init__(
        self,
        dist_class,
        model,
        value_targets,
        advantages,
        actions,
        prev_logits,
        prev_actions_logp,
        vf_preds,
        curr_action_dist,
        value_fn,
        cur_kl_coeff,
        valid_mask,
        entropy_coeff=0,
        clip_param=0.1,
        vf_clip_param=0.1,
        vf_loss_coeff=1.0,
        use_gae=True,
    ):
        """Constructs the loss for Proximal Policy Objective.
        Arguments:
            dist_class: action distribution class for logits.
            value_targets (Placeholder): Placeholder for target values; used
                for GAE.
            actions (Placeholder): Placeholder for actions taken
                from previous model evaluation.
            advantages (Placeholder): Placeholder for calculated advantages
                from previous model evaluation.
            prev_logits (Placeholder): Placeholder for logits output from
                previous model evaluation.
            prev_actions_logp (Placeholder): Placeholder for action prob output
                from the previous (before update) Model evaluation.
            vf_preds (Placeholder): Placeholder for value function output
                from the previous (before update) Model evaluation.
            curr_action_dist (ActionDistribution): ActionDistribution
                of the current model.
            value_fn (Tensor): Current value function output Tensor.
            cur_kl_coeff (Variable): Variable holding the current PPO KL
                coefficient.
            valid_mask (Optional[tf.Tensor]): An optional bool mask of valid
                input elements (for max-len padded sequences (RNNs)).
            entropy_coeff (float): Coefficient of the entropy regularizer.
            clip_param (float): Clip parameter
            vf_clip_param (float): Clip parameter for the value function
            vf_loss_coeff (float): Coefficient of the value function loss
            use_gae (bool): If true, use the Generalized Advantage Estimator.
        """
        if valid_mask is not None:

            def reduce_mean_valid(t):
                return tf.reduce_mean(tf.boolean_mask(t, valid_mask))

        else:

            def reduce_mean_valid(t):
                return tf.reduce_mean(t)

        prev_dist = dist_class(prev_logits, model)
        # Make loss functions.
        logp_ratio = tf.exp(curr_action_dist.logp(actions) - prev_actions_logp)
        action_kl = prev_dist.kl(curr_action_dist)
        self.mean_kl = reduce_mean_valid(action_kl)

        curr_entropy = curr_action_dist.entropy()
        self.mean_entropy = reduce_mean_valid(curr_entropy)

        surrogate_loss = tf.minimum(
            advantages * logp_ratio,
            advantages * tf.clip_by_value(logp_ratio, 1 - clip_param, 1 + clip_param),
        )
        self.mean_policy_loss = reduce_mean_valid(-surrogate_loss)

        # kl_loss = ... TODO
        # self.mean_kl_loss = reduce_mean_valid(kl_loss) TODO

        if use_gae:
            vf_loss1 = tf.square(value_fn - value_targets)
            vf_clipped = vf_preds + tf.clip_by_value(value_fn - vf_preds, -vf_clip_param, vf_clip_param)
            vf_loss2 = tf.square(vf_clipped - value_targets)
            vf_loss = tf.maximum(vf_loss1, vf_loss2)
            self.mean_vf_loss = reduce_mean_valid(vf_loss)
            loss = reduce_mean_valid(
                -surrogate_loss
                + cur_kl_coeff * action_kl
                + vf_loss_coeff * vf_loss
                - entropy_coeff * curr_entropy
                # + kl_loss TODO
            )
        else:
            self.mean_vf_loss = tf.constant(0.0)
            loss = reduce_mean_valid(
                -surrogate_loss
                + cur_kl_coeff * action_kl
                - entropy_coeff * curr_entropy  # + kl_loss TODO
            )
        self.loss = loss


def custom_ppo_surrogate_loss(policy, model, dist_class, train_batch):
    logits, state = model.from_batch(train_batch)
    action_dist = dist_class(logits, model)
    mask = None
    if state:
        max_seq_len = tf.reduce_max(train_batch["seq_lens"])
        mask = tf.sequence_mask(train_batch["seq_lens"], max_seq_len)
        mask = tf.reshape(mask, [-1])

    policy.loss_obj = CustomPPOLoss(
        dist_class,
        model,
        train_batch[Postprocessing.VALUE_TARGETS],
        train_batch[Postprocessing.ADVANTAGES],
        train_batch[SampleBatch.ACTIONS],
        train_batch[SampleBatch.ACTION_DIST_INPUTS],
        train_batch[SampleBatch.ACTION_LOGP],
        train_batch[SampleBatch.VF_PREDS],
        action_dist,
        model.value_function(),
        policy.kl_coeff,
        mask,
        entropy_coeff=policy.entropy_coeff,
        clip_param=policy.config["clip_param"],
        vf_clip_param=policy.config["vf_clip_param"],
        vf_loss_coeff=policy.config["vf_loss_coeff"],
        use_gae=policy.config["use_gae"],
    )
    assert(False)
    return policy.loss_obj.loss


CustomPPOTFPolicy = PPOTFPolicy.with_updates(name="CustomPPOTFPolicy", loss_fn=custom_ppo_surrogate_loss)


from ray import tune
from ray.rllib.agents.ppo.ppo import PPOTrainer


CustomPPOTrainer = PPOTrainer.with_updates(default_policy=CustomPPOTFPolicy)

tune.run(CustomPPOTrainer, config={"env": "CartPole-v0"})

```

**Comment by sven1977** [comment]
I admit, this is confusing. We will have to fix this.
For now: Can you set `get_policy_class=None`, like so:

```
CustomPPOTrainer = PPOTrainer.with_updates(default_policy=CustomPPOTFPolicy, get_policy_class=None)
```
?

**Comment by sven1977** [comment]
This fixes the issue.
Alternatively:
```
CustomPPOTrainer = PPOTrainer.with_updates(get_policy_class=lambda config: CustomPPOTFPolicy)
```

**Comment by sven1977** [comment]
Closing this issue. Feel free to re-open in case this did not work for you.
